### PR TITLE
Adding popular custom network integration

### LINKED ISSUE #12903: Create a new page: "Add custom network" in the full screen view
US: As a user I want to be able to add a popular custom network from the custom network switcher modal.
Figma link: https://www.figma.com/file/wfOwLmE52OKY0bSrsQwHPL/Add-popular-custom-network?node-id=1%3A3

Task: Create a new page: "Add custom network" in the full screen view

Acceptance criteria:
- The page according to the figma design:
image

- Networks should be listed in alphabetic order
- List of networks are grouped, e.g. groups are "Custom networks", Test networks"
- "Add button" (will open another page in the flow)
- "Add a network manually" option will open existing page in a full view, the page is shown below:
image

Open questions:
- Should be Search/Filter option be implemented like in the "pop-up" MM extension mode

### PR DESCRIPTION
**Pull request by filipsekulic**
## Explanation

Created a new feature for adding the popular custom networks. When enabled from the `Experimental` tab, this feature allows users to add one of the popular custom networks like Arbitrum One, Polygon Mainnet etc.

## More Information

Fixes: https://github.com/MetaMask/metamask-extension/issues/12903

## Screenshots/Screencaps

https://user-images.githubusercontent.com/92310504/175255384-cd9db04d-d608-4f9f-86be-5ac45eece225.mov

## Manual Testing Steps

1. Go to the `Settings`.
2. Click on the `Experimental` tab.
3. Enable the toggle switch `Show Custom Network List`.
4. From this step there are two ways to continue the flow:
    - The first one is to click on the `Networks` tab within the `Settings` and then to click on the button `Add a network`.
    - The second one is to open the network dropdown menu from the app header and then to click on the button `Add Network`.
5. The new screen is now shown. The modal view has the `back (<)` button which takes the user to the `Networks` screen. Also, the networks which rely on third parties have the exclamation triangle next to their name which can be hovered and then the tooltip with information shows up with a link `Learn more` which takes the user to the web-page `The risks of connecting to an unknown network`.
**NOTE**: If the button `Add a network manually` from the app footer is clicked, then the screen for manually adding the network is shown and from there the network can be added by entering the appropriate information about the network.
6.  The popular custom network can be added by clicking on the `Add` button. When clicked, the popup with the network information shows up. This popup has `tooltips (i)` which show information when hovered. Clicking on the `scams and network security risks` takes the user to the web-page `User Guide: Custom networks and sidechains`. If a button `View all details` is clicked then the new, smaller popup shows up with the information about the network and this popup can be closed by simply clicking on the `Close` button or `X`. 
7. The main popup can be closed by clicking on the `Cancel` button which rejects the pending approval for adding the network and the user stays on the same page. If the `Approve` button is clicked, then the network will be added, the user will be taken to the `Home page` and the new popup will show up.
8. From this step, user can switch to the newly created network by clicking on the button `Switch to network_name` or click on the `Dismiss` button and later switch to that network.

### CODE REVIEW

**Comment by filipsekulic** [comment]
Screencast:
https://user-images.githubusercontent.com/92310504/168250722-9826a290-ede6-4e5d-b203-251dd4afe929.mov

File: app/_locales/en/messages.json
Comment (darkwing) [inline app/_locales/en/messages.json:4049]: We need a `description` key here to describe what `$1` and `$2` represent.

File: ui/components/app/add-network/index.scss
Comment (darkwing) [inline ui/components/app/add-network/index.scss:80]: @georgewrmarshall Can we use this color or should it have a different named color here?

File: ui/components/app/add-network/index.scss
Comment (darkwing) [inline ui/components/app/add-network/index.scss:90]: There are a lot of `!important` usages in this file.  Can these be avoided with a parent `className`?

File: ui/components/ui/chip/chip.js
Comment (darkwing) [inline ui/components/ui/chip/chip.js:121]: Should we default `leftIconUrl` to `''`?

**Comment by darkwing** [comment]
This is how it looks in `rtl` direction:
RTLNetworks

I think some of the CSS added in this PR could be changed from `margin-left` to `margin-inline-start`, for example, to improve RTL support.  Might be worth looking at!

File: ui/components/app/add-network/index.scss
Comment (georgewrmarshall) [reply ui/components/app/add-network/index.scss:80]: Should be using one of the border color css variables. I would suggest border-muted
```suggestion
    border: 1px solid var(--color-border-muted);
```

Screen Shot 2022-05-14 at 11 32 10 AM

**Review by georgewrmarshall — COMMENTED** [review]
Looking really good nice work using the `8px` grid values for bounding box values(margin, padding). Made a couple small suggestions.

Not sure if this is a biggy? Some warning appears about the history stack when a network is deleted

https://user-images.githubusercontent.com/8112138/168446662-eb9bd5a1-88bd-4ffd-acd3-93197cdaf2a0.mov

This might be an existing prop type error but it would be good to fix: 

Screen Shot 2022-05-14 at 11 43 25 AM

File: ui/components/app/add-network/index.scss
Comment (filipsekulic) [reply ui/components/app/add-network/index.scss]: Fixed in: dc67e851ff3027e78e329448e64b8f8310629bf8

File: ui/components/app/add-network/index.scss
Comment (filipsekulic) [reply ui/components/app/add-network/index.scss]: Fixed in: dc67e851ff3027e78e329448e64b8f8310629bf8

File: ui/components/ui/chip/chip.js
Comment (filipsekulic) [reply ui/components/ui/chip/chip.js:121]: Fixed in: dc67e851ff3027e78e329448e64b8f8310629bf8

File: ui/components/app/add-network/add-network.js
Comment (filipsekulic) [reply ui/components/app/add-network/add-network.js]: Fixed in: dc67e851ff3027e78e329448e64b8f8310629bf8

File: ui/components/app/add-network/add-network.js
Comment (filipsekulic) [reply ui/components/app/add-network/add-network.js]: Fixed in: dc67e851ff3027e78e329448e64b8f8310629bf8

File: ui/components/app/add-network/add-network.js
Comment (filipsekulic) [reply ui/components/app/add-network/add-network.js]: Fixed in: dc67e851ff3027e78e329448e64b8f8310629bf8

File: ui/components/app/add-network/add-network.js
Comment (georgewrmarshall) [reply ui/components/app/add-network/add-network.js]: Still looks like it's using the wrong text color here

Screen Shot 2022-05-16 at 2 07 30 PM

Comment (filipsekulic) [reply ui/components/app/add-network/add-network.js]: Good catch! 💯  
Thanks! 😊  
I have not seen there was the `OS` theme option, and that was the part I was missing...
You can check it again and see the changes: 898658117f5fe3554764e1959a6801ea2f01b393

Comment (georgewrmarshall) [reply ui/components/app/add-network/add-network.js]: Nice one @filipsekulic! 🙏

File: ui/pages/confirmation/templates/add-ethereum-chain.js
Comment (mrbaguvix) [inline ui/pages/confirmation/templates/add-ethereum-chain.js:118]: You can reuse Line 88

File: ui/pages/confirmation/templates/add-ethereum-chain.js
Comment (mrbaguvix) [inline ui/pages/confirmation/templates/add-ethereum-chain.js:203]: Can this URL be put in a constant mapping file

**Review by 0ex-d — CHANGES_REQUESTED** [review]

**Review by georgewrmarshall — CHANGES_REQUESTED** [review]
Can you rebase from `develop` please @filipsekulic? There are some UI updates that might affect this page

Also can we make this header the same as this one so it's consistent?

This header:
Screen Shot 2022-05-26 at 1 43 16 PM

should look like this:
Screen Shot 2022-05-26 at 1 41 37 PM

Then this header should probably be updated
Screen Shot 2022-05-26 at 1 46 18 PM

**Comment by georgewrmarshall** [comment]
Also noticed the token icons get squashed when the network name is longer
Screen Shot 2022-05-26 at 12 06 49 PM

**Comment by filipsekulic** [comment]
> Can you rebase from `develop` please @filipsekulic? There are some UI updates that might affect this page
> 
> Also can we make this header the same as this one so it's consistent?
> 
> This header: Screen Shot 2022-05-26 at 1 43 16 PM
> 
> should look like this: Screen Shot 2022-05-26 at 1 41 37 PM
> 
> Then this header should probably be updated Screen Shot 2022-05-26 at 1 46 18 PM

@georgewrmarshall We discussed this on Slack and agreed to do it as it is currently being shown. (https://consensys.slack.com/archives/C02GH833686/p1651147787081969)
@bschorchit What do you think, should we leave it as it is or change it like George suggested?

**Comment by filipsekulic** [comment]
> Also noticed the token icons get squashed when the network name is longer Screen Shot 2022-05-26 at 12 06 49 PM

@georgewrmarshall Fixed:
Screenshot 2022-05-27 at 13 26 58

**Comment by Nana12345678910** [comment]
Face reigniting won't work

File: shared/constants/network.js
Comment (georgewrmarshall) [inline shared/constants/network.js:78]: We now have 2 images for BSC. See line 70. We should consolidate them and remove one

File: shared/constants/network.js
Comment (georgewrmarshall) [inline shared/constants/network.js:331]: I think we will want to reference the same constants where possible. We have defined constants at the top of this file

https://github.com/MetaMask/metamask-extension/blob/9dc94f99bea46acd888658b2e5e9ca6a3d30e438/shared/constants/network.js#L24-L30

Similarly there is also tickers and nicknames which should remain consistent across the app
https://github.com/MetaMask/metamask-extension/blob/9dc94f99bea46acd888658b2e5e9ca6a3d30e438/shared/constants/network.js#L44-L46

Could we follow this convention? Then I think the network image will show up here.

Screen Shot 2022-06-01 at 3 26 15 PM

**Review by georgewrmarshall — COMMENTED** [review]
Looking really good! A couple more suggestions

File: shared/constants/network.js
Comment (filipsekulic) [reply shared/constants/network.js]: Fixed: f0c4ca1d42e706eb710e1c2d54578a354d2de0ee

File: shared/constants/network.js
Comment (filipsekulic) [reply shared/constants/network.js]: Fixed: f0c4ca1d42e706eb710e1c2d54578a354d2de0ee

Screenshot 2022-06-02 at 13 01 45

**Review by georgewrmarshall — DISMISSED** [review]
Excellent work! Looks like there is one conflict that needs to be resolved but UI looks great to me! 💯 I have a few other UI updates that touch network but I'll save them for another PR.

**Comment by filipsekulic** [comment]
> Excellent work! Looks like there is one conflict that needs to be resolved but UI looks great to me! 💯 I have a few other UI updates that touch network but I'll save them for another PR.

Hey, @georgewrmarshall! 
Fixed it. 🙂

**Review by georgewrmarshall — DISMISSED** [review]

**Comment by seaona** [comment]
Great feature! I've noticed that when we Cancel the Add Network process we get a console error `stack": "Error: User rejected the request` and we land to Home page instead of staying on the Network page.

https://user-images.githubusercontent.com/54408225/172439996-65d280b9-7bca-4d9e-aea0-1559ac7a4ce2.mp4

Could you double-check if that's okay @filipsekulic?

**Comment by seaona** [comment]
Looks like all the Popular Networks that point to Infura are failing their requests and no data is fetched. I also see this error on the background console: `project ID does not have access to XYZ network`. Affected networks:
- Arbitrum one
- Polygon
- Optimism

grafik


https://user-images.githubusercontent.com/54408225/172581933-a2021982-011b-42ea-997f-7d1789c7766b.mp4

File: ui/components/app/dropdowns/network-dropdown.js
Comment (danjm) [reply ui/components/app/dropdowns/network-dropdown.js:135]: That's true. @filipsekulic I guess the if/else can be removed here?

**Comment by danjm** [comment]
We make the "Add Popular Network" screen only accessible if the user toggles on a setting in "Experimental". The same as is currently done for "User token detection" and "Enable Enhanced Gas Fee UI". That toggle should control a feature flag the same as the token detection and enhanced gas fee UI toggles do. Then, all that setting needs to do is direct the user to the `ADD_NETWORKS_ROUTE` everywhere there is currently a `history.push` to the `ADD_POPULAR_NETWORKS_ROUTE`

Some something like `history.push(addPopularNetworkFeatureToggledOn ? ADD_POPULAR_CUSTOM_NETWORK : ADD_NETWORK_ROUTE);`

We just need this to be in place for one or two releases after we first release the feature, and then we can turn the toggle on by default for everyone.

I'd suggest addressing the issues found by @seaona first, and then adding this toggle

fyi @bschorchit

File: ui/components/app/add-network/add-network.js
Comment (darkwing) [inline ui/components/app/add-network/add-network.js:69]: This theme stuff seems out of place -- does it need to be here?

**Review by darkwing — CHANGES_REQUESTED** [review]

File: app/scripts/metamask-controller.js
Comment (filipsekulic) [reply app/scripts/metamask-controller.js:2052]: Fixed here: b9688b9f1eedc11dee8d27d9ff500555cb008b9f

File: ui/components/app/add-network/add-network.js
Comment (filipsekulic) [reply ui/components/app/add-network/add-network.js:69]: Yes, we need it here. If a theme is `os` then we have to find out which one to use (light or dark). That is being checked here.

**Comment by filipsekulic** [comment]
@seaona @danjm @georgewrmarshall @darkwing 
Hey! Just wanted to notify you that I updated the PR. You can also check the toggle switch within the experimental tab for this new feature.

File: ui/components/app/add-network/add-network.js
Comment (danjm) [reply ui/components/app/add-network/add-network.js:69]: @darkwing This is to pass the correct theme to a tooltip component below. Is there something different that should happen if `theme === THEME_TYPE.OS`? How is that case handled in the rest of the app?

Comment (darkwing) [reply ui/components/app/add-network/add-network.js:69]: OK, I see that now; I'm wondering if we should move this theme check to the actual Tooltip component itself?  I'm also wondering if there's any advantage to setting the theme in local state.

**Comment by seaona** [comment]
I see all the reported bugs fixed now. Looking good 🙌
Just a small observation: since now we've introduced the toggle button under experimental, should we also include this new section on Settings Search? @filipsekulic @danjm

image

**Comment by seaona** [comment]
@georgewrmarshall should we capitalize any letter of the `Show custom network list`? I'm not sure if there is a convention as sometimes I see capital letters and in some cases don't. Ty!

image

image

**Comment by georgewrmarshall** [comment]
> @georgewrmarshall should we capitalize any letter of the `Show custom network list`? I'm not sure if there is a convention as sometimes I see capital letters and in some cases don't. Ty!

@seaona great question! I've reached out to Corey our content designer for some feedback. It looks like most are title case so I would tentatively push for the most common casing. We should probably do a complete overhaul of our content file in `app/_locales/en/messages.json` at some stage I've shared that with Corey too.

**Review by georgewrmarshall — COMMENTED** [review]
Looking really good! Not sure if it's a caching issue but I get the loading spinner for my balance for Arbitrum One and Optimism networks.

Screen Shot 2022-06-20 at 11 49 57 AM
Screen Shot 2022-06-20 at 11 52 09 AM

File: ui/components/app/add-network/add-network.js
Comment (georgewrmarshall) [reply ui/components/app/add-network/add-network.js:69]: Created a PR to fix this https://github.com/MetaMask/metamask-extension/pull/14992
Theme logic shouldn't be required once this has been merged to develop

**Comment by seaona** [comment]
thank you v much @filipsekulic. Just checked and for my side, everything looks great 🙌 incredible job @filipsekulic!

File: ui/components/app/add-network/add-network.js
Comment (georgewrmarshall) [inline ui/components/app/add-network/add-network.js:22]: Can remove with new Tooltip update https://github.com/MetaMask/metamask-extension/pull/14992
```suggestion
```

File: ui/components/app/add-network/add-network.js
Comment (georgewrmarshall) [inline ui/components/app/add-network/add-network.js:69]: Can remove with Tooltip update https://github.com/MetaMask/metamask-extension/pull/14992

```suggestion
```

**Review by georgewrmarshall — CHANGES_REQUESTED** [review]
Awesome! Looks and works great. I managed to [sneak an update to the Tooltip into `develop`](https://github.com/MetaMask/metamask-extension/pull/14992) so there isn't any need to pass a theme to it anymore. Have made the relevant code suggestions.

Checked it locally on your branch and it seems to work. Will approve after these updates have been made 🙏 

Screen Shot 2022-06-21 at 11 48 56 AM
Screen Shot 2022-06-21 at 11 48 06 AM
Screen Shot 2022-06-21 at 11 48 32 AM

File: ui/components/app/add-network/add-network.js
Comment (georgewrmarshall) [inline ui/components/app/add-network/add-network.js:61]: Can remove with Tooltip update https://github.com/MetaMask/metamask-extension/pull/14992

```suggestion
```

**Comment by filipsekulic** [comment]
@georgewrmarshall 
Nice! 
I made those changes here: 57f24d94352b8159b6df948c01add9997efae289

**Review by georgewrmarshall — APPROVED** [review]
UI LGTM! Nice work 💯 Could you please add a PR description and "Manual Testing Steps" it helps the [support content folks write documentation](https://consensys.slack.com/archives/G1L7H42BT/p1655912991964999)

- verified adding / removing networks
- verified UI in light/dark mode
- verified responsive layout
- verified flow in popover/action view

**Comment by filipsekulic** [comment]
Hey @georgewrmarshall! Thank you!
You've got the point! I updated the PR description.

**Comment by coreyjanssen** [comment]
@seaona @georgewrmarshall re: Capitalization, we should be using sentence case throughout

**Review by danjm — APPROVED** [review]

**Review by darkwing — APPROVED** [review]
Approved.

From a UX perspective, I wish we didn't send the user home, in the case that they want to add a handful of networks, but this is a huge new feature we should get out there ASAP.   Well done!

**Pull request merged**
